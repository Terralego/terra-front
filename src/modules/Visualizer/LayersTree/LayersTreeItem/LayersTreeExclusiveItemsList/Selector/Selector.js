import React from 'react';
import PropTypes from 'prop-types';
import { Callout } from '@blueprintjs/core/lib/cjs/components/callout/callout';
import { Intent } from '@blueprintjs/core/lib/cjs/common/intent';

import Select from '../../../../../Forms/Controls/Select';
import translateMock from '../../../../../../utils/translate';

import './styles.scss';

const SelectorSelect = ({
  label,
  name,
  values,
  currentValue,
  onChange,
  ...props
}) => {
  const handleChange = React.useCallback(
    value => onChange(name, value),
    [name, onChange],
  );

  /**
   * Get selected value from all values.
   */
  const getSelectedValue = React.useMemo(() => {
    if (values.length) {
      const { value } = values.find(({ value: listValue }) =>
        currentValue === listValue) || values[0];
      return value;
    }
    return null;
  }, [currentValue, values]);

  return (
    <Select
      fullWidth
      label={label}
      className="selector__select"
      onChange={handleChange}
      values={values}
      value={getSelectedValue}
      {...props}
    />
  );
};


export const Selector = ({
  selectors,
  layers,
  activeLayer: { selectorKey = {} } = {},
  onChange,
  translate,
}) => {
  const [noMatchingLayer, setNoMatchingLayer] = React.useState(false);

  const getLayerForSelector = React.useCallback(currentSelectorKey =>
    layers.findIndex(({ selectorKey: layerSelectorKey }) =>
      Object.keys(layerSelectorKey).length === selectors.length &&
      selectors.every(({ name: selectorName }) =>
        layerSelectorKey[selectorName] === currentSelectorKey[selectorName])),
  [layers, selectors]);

  const handleChange = React.useCallback((name, value) => {
    const newKey = { ...selectorKey, [name]: value };

    const newActiveLayerIndex = getLayerForSelector(newKey);

    if (newActiveLayerIndex === -1) {
      setNoMatchingLayer(true);
      return;
    }

    setNoMatchingLayer(false);
    onChange(newActiveLayerIndex);
  }, [getLayerForSelector, onChange, selectorKey]);

  React.useEffect(() => {
    const activeLayerIndex = getLayerForSelector(selectorKey);
    setNoMatchingLayer(activeLayerIndex === -1);
  }, [getLayerForSelector, selectorKey]);

  return (
    <div className="selector">
      {selectors.map(({ label, name, values }) => (
        <SelectorSelect
          key={name}
          name={name}
          label={label}
          onChange={handleChange}
          values={values}
          currentValue={selectorKey[name]}
        />
      ))}
      {noMatchingLayer && (
      <Callout intent={Intent.WARNING} className="selector__message">
        {translate('terralego.visualizer.layerstree.group.selector')}
      </Callout>
      )}
    </div>
  );
};

Selector.propTypes = {
  selectors: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    name: PropTypes.string.isRequired,
    values: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })).isRequired,
  })).isRequired,
  layers: PropTypes.arrayOf(PropTypes.shape({
    selectorKey: PropTypes.object,
  })).isRequired,
  activeLayer: PropTypes.shape({
    selectorKey: PropTypes.object,
  }).isRequired,
  onChange: PropTypes.func,
  translate: PropTypes.func,
};

Selector.defaultProps = {
  onChange () {},
  translate: translateMock({
    'terralego.visualizer.layerstree.group.selector': 'No layer found',
  }),
};

export default Selector;
