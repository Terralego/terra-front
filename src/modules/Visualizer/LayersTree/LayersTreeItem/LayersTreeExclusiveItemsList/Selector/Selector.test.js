import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';

import Selector from './Selector';

jest.mock('uuid/v4', () => () => 'uuid');

it('should render', () => {
  const { asFragment } = render(
    <Selector
      selectors={[{
        label: 'Foo',
        name: 'foo',
        values: [{ label: 'Foo', value: 'foo' }, { label: 'Bar', value: 'bar' }],
      }, {
        label: 'Bar',
        name: 'bar',
        values: [{ label: 'Foo', value: 'foo' }, { label: 'Bar', value: 'bar' }],
      }]}
      layers={[{
        label: 'Layer 1',
        selectorKey: {
          foo: 'foo',
          bar: 'foo',
        },
      }, {
        label: 'Layer 1',
        selectorKey: {
          foo: 'foo',
          bar: 'bar',
        },
      }]}
      activeLayer={{
        selectorKey: {
          foo: 'foo',
          bar: 'foo',
        },
      }}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

it('should render even without selector', () => {
  const { asFragment } = render(
    <Selector
      selectors={[{
        label: 'Foo',
        name: 'foo',
        values: [{ label: 'Foo', value: 'foo' }, { label: 'Bar', value: 'bar' }],
      }, {
        label: 'Bar',
        name: 'bar',
        values: [{ label: 'Foo', value: 'foo' }, { label: 'Bar', value: 'bar' }],
      }]}
      layers={[{
        label: 'Layer 1',
        selectorKey: {
          foo: 'foo',
          bar: 'foo',
        },
      }, {
        label: 'Layer 1',
        selectorKey: {
          foo: 'foo',
          bar: 'bar',
        },
      }]}
      activeLayer={{}}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

it('should render an error message', async () => {
  const { asFragment } = render(
    <Selector
      selectors={[{
        label: 'Foo',
        name: 'foo',
        values: [{ label: 'Foo1', value: 'foo' }, { label: 'Bar1', value: 'bar' }],
      }, {
        label: 'Bar',
        name: 'bar',
        values: [{ label: 'Foo2', value: 'foo' }, { label: 'Bar2', value: 'bar' }],
      }]}
      layers={[{
        label: 'Layer 1',
        selectorKey: {
          foo: 'foo',
          bar: 'foo',
        },
      }, {
        label: 'Layer 1',
        selectorKey: {
          foo: 'foo',
          bar: 'bar',
        },
      }]}
      activeLayer={{
        selectorKey: {
          foo: 'bar', // Layer with this selector don't exists
          bar: 'foo',
        },
      }}
    />,
  );

  expect(asFragment()).toMatchSnapshot();
});

it('should change layer', async () => {
  const selectors = [{
    label: 'Foo',
    name: 'foo',
    values: [{ label: 'Foo1', value: 'foo' }, { label: 'Bar1', value: 'bar' }],
  }, {
    label: 'Bar',
    name: 'bar',
    values: [{ label: 'Foo2', value: 'foo' }, { label: 'Bar2', value: 'bar' }, { label: 'Baz2', value: 'baz' }],
  }];
  const layers = [{
    label: 'Layer 1',
    selectorKey: {
      foo: 'foo',
      bar: 'foo',
    },
  }, {
    label: 'Layer 2',
    selectorKey: {
      foo: 'foo',
      bar: 'bar',
    },
  }, {
    label: 'Layer 3',
    selectorKey: {
      foo: 'bar',
      bar: 'foo',
    },
  }, {
    label: 'Layer 4',
    selectorKey: {
      foo: 'bar',
      bar: 'bar',
    },
  }];
  const onChange = jest.fn();

  const { asFragment } = render(
    <Selector
      selectors={selectors}
      layers={layers}
      activeLayer={layers[0]}
      onChange={onChange}
    />,
  );

  // First selector
  fireEvent.click(screen.getByText('Foo1'));
  fireEvent.click(screen.getByText('Bar1'));

  expect(onChange).toHaveBeenLastCalledWith(2);

  onChange.mockClear();

  // Second selector
  fireEvent.click(screen.getByText('Foo2'));

  fireEvent.click(screen.getByText('Bar2'));

  expect(onChange).toHaveBeenLastCalledWith(1);

  onChange.mockClear();

  fireEvent.click(screen.getByText('Baz2')); // Selector with no layer

  expect(onChange).not.toHaveBeenCalled();

  expect(asFragment()).toMatchSnapshot();
});
