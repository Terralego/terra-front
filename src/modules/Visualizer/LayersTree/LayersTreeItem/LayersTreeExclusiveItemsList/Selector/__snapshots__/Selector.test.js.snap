// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change layer 1`] = `
<DocumentFragment>
  <div
    class="selector"
  >
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Foo
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Foo1
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
        <div
          class="bp3-overlay bp3-overlay-inline"
        >
          <div
            class="bp3-transition-container bp3-popover-exit bp3-popover-exit-active"
            style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
          >
            <div
              class="bp3-popover bp3-minimal bp3-select-popover"
            >
              <div
                class="bp3-popover-content"
              >
                <div>
                  <ul
                    class="bp3-menu"
                  >
                    <div
                      class="control-container__item"
                    >
                      <li
                        class=""
                      >
                        <a
                          class="bp3-menu-item bp3-active bp3-intent-primary bp3-popover-dismiss"
                        >
                          <div
                            class="bp3-text-overflow-ellipsis bp3-fill"
                          >
                            Foo1
                          </div>
                        </a>
                      </li>
                    </div>
                    <div
                      class="control-container__item"
                    >
                      <li
                        class=""
                      >
                        <a
                          class="bp3-menu-item bp3-popover-dismiss"
                        >
                          <div
                            class="bp3-text-overflow-ellipsis bp3-fill"
                          >
                            Bar1
                          </div>
                        </a>
                      </li>
                    </div>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Bar
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Foo2
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
        <div
          class="bp3-overlay bp3-overlay-inline"
        >
          <div
            class="bp3-transition-container bp3-popover-exit bp3-popover-exit-active"
            style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
          >
            <div
              class="bp3-popover bp3-minimal bp3-select-popover"
            >
              <div
                class="bp3-popover-content"
              >
                <div>
                  <ul
                    class="bp3-menu"
                  >
                    <div
                      class="control-container__item"
                    >
                      <li
                        class=""
                      >
                        <a
                          class="bp3-menu-item bp3-active bp3-intent-primary bp3-popover-dismiss"
                        >
                          <div
                            class="bp3-text-overflow-ellipsis bp3-fill"
                          >
                            Foo2
                          </div>
                        </a>
                      </li>
                    </div>
                    <div
                      class="control-container__item"
                    >
                      <li
                        class=""
                      >
                        <a
                          class="bp3-menu-item bp3-popover-dismiss"
                        >
                          <div
                            class="bp3-text-overflow-ellipsis bp3-fill"
                          >
                            Bar2
                          </div>
                        </a>
                      </li>
                    </div>
                    <div
                      class="control-container__item"
                    >
                      <li
                        class=""
                      >
                        <a
                          class="bp3-menu-item bp3-popover-dismiss"
                        >
                          <div
                            class="bp3-text-overflow-ellipsis bp3-fill"
                          >
                            Baz2
                          </div>
                        </a>
                      </li>
                    </div>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
    <div
      class="bp3-callout bp3-intent-warning bp3-callout-icon selector__message"
    >
      <span
        class="bp3-icon bp3-icon-warning-sign"
        icon="warning-sign"
      >
        <svg
          data-icon="warning-sign"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <desc>
            warning-sign
          </desc>
          <path
            d="M19.86 17.52l.01-.01-9-16-.01.01C10.69 1.21 10.37 1 10 1s-.69.21-.86.52l-.01-.01-9 16 .01.01c-.08.14-.14.3-.14.48 0 .55.45 1 1 1h18c.55 0 1-.45 1-1 0-.18-.06-.34-.14-.48zM11 17H9v-2h2v2zm0-3H9V6h2v8z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      No layer found
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render 1`] = `
<DocumentFragment>
  <div
    class="selector"
  >
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Foo
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Foo
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
      </span>
    </div>
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Bar
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Foo
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render an error message 1`] = `
<DocumentFragment>
  <div
    class="selector"
  >
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Foo
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Bar1
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
      </span>
    </div>
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Bar
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Foo2
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
      </span>
    </div>
    <div
      class="bp3-callout bp3-intent-warning bp3-callout-icon selector__message"
    >
      <span
        class="bp3-icon bp3-icon-warning-sign"
        icon="warning-sign"
      >
        <svg
          data-icon="warning-sign"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <desc>
            warning-sign
          </desc>
          <path
            d="M19.86 17.52l.01-.01-9-16-.01.01C10.69 1.21 10.37 1 10 1s-.69.21-.86.52l-.01-.01-9 16 .01.01c-.08.14-.14.3-.14.48 0 .55.45 1 1 1h18c.55 0 1-.45 1-1 0-.18-.06-.34-.14-.48zM11 17H9v-2h2v2zm0-3H9V6h2v8z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      No layer found
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render even without selector 1`] = `
<DocumentFragment>
  <div
    class="selector"
  >
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Foo
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Foo
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
      </span>
    </div>
    <div
      class="control-container control-container--select control-container--full-width"
      role="presentation"
    >
      <label
        class="control-label"
        for="select-uuid"
      >
        Bar
      </label>
      <span
        class="bp3-popover-wrapper tf-select selector__select"
      >
        <span
          class="bp3-popover-target"
        >
          <div
            class=""
          >
            <button
              class="bp3-button"
              id="select-uuid"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Foo
              </span>
              <span
                class="bp3-icon bp3-icon-double-caret-vertical"
                icon="double-caret-vertical"
              >
                <svg
                  data-icon="double-caret-vertical"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    double-caret-vertical
                  </desc>
                  <path
                    d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </span>
      </span>
    </div>
    <div
      class="bp3-callout bp3-intent-warning bp3-callout-icon selector__message"
    >
      <span
        class="bp3-icon bp3-icon-warning-sign"
        icon="warning-sign"
      >
        <svg
          data-icon="warning-sign"
          height="20"
          viewBox="0 0 20 20"
          width="20"
        >
          <desc>
            warning-sign
          </desc>
          <path
            d="M19.86 17.52l.01-.01-9-16-.01.01C10.69 1.21 10.37 1 10 1s-.69.21-.86.52l-.01-.01-9 16 .01.01c-.08.14-.14.3-.14.48 0 .55.45 1 1 1h18c.55 0 1-.45 1-1 0-.18-.06-.34-.14-.48zM11 17H9v-2h2v2zm0-3H9V6h2v8z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      No layer found
    </div>
  </div>
</DocumentFragment>
`;
